### PR TITLE
[MPS] Add CompileShader method

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141478

This allows one to do something like that
```python
import torch
x = torch.ones(10, device="mps")
m = torch.mps._compile_shader("""
   kernel void foo(device float* x, uint idx [[thread_position_in_grid]]) {
     x[idx] += idx;
   }
")
m.foo(x)
```

And in general enables writing custom operators using Metal shaders purely in Python

### CODE REVIEW

**Comment by qqaatw** [comment]
I'm curious about the use of this addition since it accepts one argument only. I assume there is no cuda counterpart in PyTorch.

**Comment by malfet** [comment]
> I'm curious about the use of this addition since it accepts one argument only. I assume there is no cuda counterpart in PyTorch.

@qqaatw it's still work in progress :) But final version will accept any number of arguments. As for usefulness, at the very least it unblocks quick prototyping and usecases like shader toys, but in the long run it can enable a pretty advanced codegen.

**Review by manuelcandales — APPROVED** [review]

File: test/test_mps.py [hunk 1]
```diff
@@ -12480,6 +12480,39 @@ def ones(device):
             cpu_tensor = ones("cpu")
             self.assertEqual(mps_tensor.cpu(), cpu_tensor)
 
+class TestMetalLibrary(TestCaseMPS):
```
[the remark this answers is not in the document]
Comment (malfet) [reply test/test_mps.py:12483]: Done

File: torch/mps/__init__.py [hunk 2]
```diff
@@ -140,6 +140,12 @@ def recommended_max_memory() -> int:
     return torch._C._mps_recommendedMaxMemory()
 
 
+def _compile_shader(source: str):
```
[the remark this answers is not in the document]
Comment (malfet) [reply torch/mps/__init__.py:143]: I think this should be done as a separate wrapper in the Python layer, where we create an abstraction (similar to TensorIterator) and developer only need to implement the function that is executed in the main loop

**Pull request closed without merging**